### PR TITLE
Refactor video processing pipeline

### PR DESCRIPTION
**Pull request by pgrill**
Refactoring the video processing pipeline and the loggers structure.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -299,27 +298,44 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
-  - `Name`: Configures the chosen logger. For now, we only support the *csv_logger*.
-  - `TimeInterval`: Sets the desired logging interval for objects detections and violations.
-  - `LogDirectory`: Defines the location where the generated files will be stored.
-  - `EnableReports`: A boolean parameter to enable/disable the reports generation.
-  - `HeatmapResolution`: Sets the resolution used by the heatmap report.
-  - `WebHooksEndpoint`: Configures an endpoint to receive in real-time the objects detections and violations.
+
+- `[SourcePostProcessor_N]`:
+
+  In the config files, we use the *SourcePostProcessor* sections to specifies additional processing steps in the sources. We support 3 different ones (identified by the field *Name*) that you enable/disable uncommenting/commenting them.
```
Comment (mats-claassen) [inline README.md:314]:
```suggestion
  In the config files, we use the *SourcePostProcessor* sections to specify additional processing steps after running the detector and face mask classifier (if available) on the video sources. We support 3 different ones (identified by the field *Name*) that you enable/disable uncommenting/commenting them.
```

File: README.md [hunk 2]
```diff
@@ -299,27 +298,44 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
+    - `social_distance`: Used to measure the distance between objects and detect social distancing violations.
+      - `DefaultDistMethod`: Defines the default distance algorithm for the cameras without *DistMethod* configuration.
+      - `DistThreshold`: Configures the distance threshold for the *social distancing violations*
+    - `anonymizer`: A step used to enable anonymization of faces in videos and screenshots.
+
+- `[SourceLogger_N]`:
+
+  Similar to the section *SourcePostProcessor_N*, we support multiple loggers (right now 4) that you enable/disable uncommenting/commenting them.
+  - `video_logger`: Generates a video stream with the processing results. It is a useful logger to monitor in real-time your sources.
+  - `s3_logger`: Store the processed data in a S3 bucket.
```
Comment (mats-claassen) [inline README.md:326]: This currently stores only screenshots. Maybe add that here in the description. In future we could store all data, and then we would need more parameters

File: README.md [hunk 3]
```diff
@@ -299,27 +298,44 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
+    - `ScreenshotPeriod`: Defines a time period (expressed in minutes) to take a screenshot of all the cameras and store them in S3. If you set the value to 0, no screenshots will be taken.
+    - `ScreenshotS3Bucket`: Configures the S3 Bucket used to store the screenshot.
+  - `file_system_logger`: Store the processed data in a folder inside the processor.
+    - `TimeInterval`: Sets the desired logging interval for objects detections and violations.
+    - `LogDirectory`: Defines the location where the generated files will be stored.
+     - `ScreenshotPeriod`: Defines a time period (expressed in minutes) to take a screenshot of all the cameras and store them. If you set the value to 0, no screenshots will be taken.
```
Comment (mats-claassen) [inline README.md:332]: This should have the same indentation as the previous line, right?

File: README.md [hunk 4]
```diff
@@ -299,27 +298,44 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
+  - `web_hook_logger`: Allows you to configure an external endpoint to receive in real-time the objects detections and violations.
```
Comment (mats-claassen) [inline README.md:333]:
```suggestion
  - `web_hook_logger`: Allows you to configure an external endpoint to receive in real-time the object detections and violations.
```

File: README.md [hunk 5]
```diff
@@ -299,27 +298,44 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
+    - `Endpoint`: Configures an endpoint url.
+ 
+- `[PeriodicTask_N]`:
+
+  The processor also supports the execution of periodic tasks to generated reports, accumulated metrics, etc. For now, we support only the *reports* task.
```
Comment (mats-claassen) [inline README.md:338]:
```suggestion
  The processor also supports the execution of periodic tasks to generate reports, accumulate metrics, etc. For now, we support only the *reports* task.
```

File: config-x86-gpu.ini [hunk 6]
```diff
@@ -73,28 +71,47 @@ ClassID = 1
...
-; Number of times tracker was lost while tracking
+[Tracker]
+Name = IOUTracker
 MaxLost = 5
 TrackerIOUThreshold = 0.5
+
+[SourcePostProcessor_0]
+Name = objects_filtering
 NMSThreshold = 0.98
-; Default distance measurement method (used when no DistMethod is specified for the source):
```
Comment (mats-claassen) [inline config-x86-gpu.ini:85]: I think the comments are useful and we should try to keep them. Same for the comments that mention which models are available

File: README.md (hunk 2 above)
Comment (pgrill) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (pgrill) [reply README.md]: Done

File: libs/loggers/video_logger.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        self.birds_eye_resolution = (200, 300)
+        self.out = None
+        self.out_birdseye = None
+        self.live_feed_enabled = live_feed_enabled
+        self.track_hist = dict()
+
+    def start_logging(self, fps):
+        if not self.live_feed_enabled:
+            return
+        self.out, self.out_birdseye = (
```
Comment (mats-claassen) [inline libs/loggers/video_logger.py:25]: Birdseye is not used at all in the dashboard. Should we remove it?

File: libs/loggers/web_hook_logger.py [hunk 8]
```diff
@@ -29,3 +36,58 @@ def log_objects(self, objects, violating_objects, violating_objects_index_list,
...
+    def update(self, cv_image, objects, post_processing_data, fps):
+        violating_objects = post_processing_data.get("violating_objects", [])
+        if not self.web_hook_endpoint:
+            return
+        if time.time() - self.submited_time > self.time_interval:
+            # Get timeline which is used for as Timestamp
+            now = datetime.now()
+            current_time = now.strftime("%Y-%m-%d %H:%M:%S")
+            # Process objects
+            objects_formated = self.format_objects(objects)
```
Comment (mats-claassen) [inline libs/loggers/web_hook_logger.py:49]: This function is very similar to the csv logger. Could we abstract more of it to a superclass? I think the format_objects function is the same

File: libs/utils/loggers.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+def get_source_log_directory(config):
```
Comment (mats-claassen) [inline libs/utils/loggers.py:1]: What about calling this file differently. Maybe logger_utils.py

File: libs/loggers/video_logger.py (hunk 7 above)
Comment (pgrill) [reply libs/loggers/video_logger.py:25]: Is not used now. However, I am not sure that we don't want to include it in the future.

File: libs/utils/loggers.py (hunk 9 above)
Comment (pgrill) [reply libs/utils/loggers.py:1]: All the functions inside the utils folder don't include the `_utils`. If we change the name for this file we should also change the other ones.

File: README.md [hunk 10]
```diff
@@ -239,13 +239,12 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
-  - `ScreenshotPeriod`: Defines a time period (expressed in minutes) to take a screenshot of all the cameras and store them in S3. If you set the value to 0, no screenshots will be taken.
-  - `ScreenshotS3Bucket`: Configures the S3 Bucket used to store the screenshot.
   - `ScreenshotsDirectory`: Configures the folder dedicated to storing all the images (for example the heatmap reports) generated by the processor. We recommend to set this folder to a mounted directory (such as */repo/data/processor/static/screenshots*)
   - `DashboardURL`: Sets the url where the frontend is running. Unless you are using a custom domain, you should keep this value as https://beta.lanthorn.ai/.
   - `SlackChannel`: Configures the slack channel used by the notifications. The chosen slack channel must exist in the configured workspace.
   - `OccupancyAlertsMinInterval`:  Sets the desired interval (in seconds) between occupancy alerts.
   - `MaxThreadRestarts`: Defines the number of restarts allowed per thread.
+  - `HeatmapResolution`: Sets the resolution used by the heatmap report.
```
[the remark this answers is not in the document]
Comment (pgrill) [reply README.md:247]: It was configured in the Logger section before (that was removed). 
Maybe make more sense to put it in the report's sections :thinking:

File: libs/area_threading.py [hunk 11]
```diff
@@ -36,25 +36,29 @@ def __init__(self, config, area):
...
+                    if (datetime.now() - last_restart_time).total_seconds() > 60:
+                        # If the last restart was previous than 1 minute ago, restart the counter.
+                        restarts = 0
+                    if restarts == max_restarts:
+                        raise e
+                    # Sleep the thread for 5 seconds and try to process the area again
+                    time.sleep(5)
+                    logging.info("Restarting the area processing")
+                    restarts += 1
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (pgrill) [reply libs/area_threading.py:59]: Yes, if you forgot to specify a mandatory field in the config you will receive an error In the engine init

File: README.md [hunk 12]
```diff
@@ -239,13 +239,12 @@ All the configurations are grouped in *sections* and some of them can vary depen
   - `Resolution`: Specifies the image resolution that the whole processor will use. If you are using a single camera we recommend using that resolution.
   - `Encoder`: Specifies the video encoder used by the processing pipeline.
   - `MaxProcesses`: Defines the number of processes executed in the processor. If you are using multiple cameras per processor we recommend increasing this number.
-  - `ScreenshotPeriod`: Defines a time period (expressed in minutes) to take a screenshot of all the cameras and store them in S3. If you set the value to 0, no screenshots will be taken.
-  - `ScreenshotS3Bucket`: Configures the S3 Bucket used to store the screenshot.
   - `ScreenshotsDirectory`: Configures the folder dedicated to storing all the images (for example the heatmap reports) generated by the processor. We recommend to set this folder to a mounted directory (such as */repo/data/processor/static/screenshots*)
```
[the remark this answers is not in the document]
Comment (pgrill) [reply README.md:242]: This folder is used not only by the logger (for example, the heatmap snapshots are also stored there).

File: README.md [hunk 13]
```diff
@@ -288,7 +287,7 @@ All the configurations are grouped in *sections* and some of them can vary depen
   - `DistMethod`: Configures the chosen distance method used by the processor to detect the violations. There are three different values: CalibratedDistance, CenterPointsDistance and FourCornerPointsDistance. If you want to use *CalibratedDistance* you will need to calibrate the camera from the [UI](https://beta.lanthorn.ai).
 
 - `[Detector]`:
-  - `Device`: Specifies the device. The available values are *Jetson*, *EdgeTPU*, *Dummy*, *x86*
+  - `Device`: Specifies the device. The available values are *Jetson*, *EdgeTPU*, *Dummy*, *x86*, *x86-gpu*
```
[the remark this answers is not in the document]
Comment (pgrill) [reply README.md:290]: For now, you are using the same device to execute all the pipeline steps. Maybe in the future, we choose to execute the classifiers in a different one. For that reason, I duplicated it.

File: config-coral.ini [hunk 14]
```diff
@@ -68,35 +66,61 @@ Device = EdgeTPU
...
-WebHooksEndpoint=
+; Screenshot time is measured in minutes (if period <= 0 then no screenshots are uploaded)
+ScreenshotPeriod = 0
+
+[SourceLogger_3]
+Name = web_hook_logger
+Endpoint =
+TimeInterval = 0.5
+
+; Enable the PeriodicTask_0 if you want to generate reports
```
[the remark this answers is not in the document]
Comment (pgrill) [reply config-coral.ini:124]: I didn't check it

File: config-coral.ini [hunk 15]
```diff
@@ -42,7 +40,7 @@ DailyReport = False
 DailyReportTime = 06:00
 
 [Source_0]
-VideoPath = /repo/data/softbio_vid.mp4
+VideoPath = /repo/data/video_demo.mkv
```
[the remark this answers is not in the document]
Comment (pgrill) [reply config-coral.ini]: No, I committed it by mistake

File: config-coral.ini [hunk 16]
```diff
@@ -68,35 +66,61 @@ Device = EdgeTPU
...
 ; - FourCornerPointsDistance: compare four corresponding points of pedestrian boxes and get the minimum of them.
 DefaultDistMethod = CenterPointsDistance
-; distance threshold for smart distancing in (cm)
 DistThreshold = 150
-Anonymize = true
 
-[Logger]
-Name = csv_logger
+[SourcePostProcessor_2]
+Name = anonymizer
```
[the remark this answers is not in the document]
Comment (pgrill) [reply config-coral.ini:101]: I committed it by mistake.

File: libs/classifiers/classifier.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    def __init__(self, config):
+        self.config = config
+        self.classifier = None
+        classifier_section = self.config.get_section_dict("Classifier")
+        if classifier_section["Device"] == "EdgeTPU":
+            from .edgetpu.classifier import Classifier as EdgeClassifier
+            self.classifier = EdgeClassifier(self.config)
+        elif classifier_section["Device"] in ["x86", "x86-gpu"]:
+            from .x86.classifier import Classifier as X86Classifier
+            self.classifier = X86Classifier(self.config)
```
[the remark this answers is not in the document]
Comment (pgrill) [reply libs/classifiers/classifier.py:12]: We don't have support for jetson yet.

File: libs/classifiers/classifier.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+        else:
+            raise ValueError(f"Not supported device named: {classifier_section['Device']}")
+        self.min_threshold = float(self.config.get_section_dict("Classifier").get("MinScore", 0.75))
+
+    def inference(self, objects):
+        return self.classifier.inference(objects)
+
+    def object_post_process(self, object, classifier_result, classifier_score):
```
[the remark this answers is not in the document]
Comment (pgrill) [reply libs/classifiers/classifier.py:20]: Yes, if you configure a valid facemask classifier but you don't have face detection in your detector (for example `mobilenet_ssd_v2`) the face_label will be `-1`.

File: config-coral.ini (hunk 15 above)
[the remark this answers is not in the document]
Comment (pgrill) [reply config-coral.ini]: I committed it by mistake

File: libs/loggers/web_hook_logger.py (hunk 8 above)
Comment (pgrill) [reply libs/loggers/web_hook_logger.py]: Yes, you are right. I created a `RawDataLogger`  parent of `FileSystemLogger` and `WebHookLogger`

File: api/cameras.py [hunk 19]
```diff
@@ -255,13 +255,17 @@ async def get_camera_calibration_image(camera_id: str):
     """
     Gets the image required to calibrate the camera <camera_id>
     """
+    logger.info("Camera id")
```
Comment (mats-claassen) [inline api/cameras.py:258]: I think this can be removed

File: libs/utils/loggers.py (hunk 9 above)
Comment (mats-claassen) [reply libs/utils/loggers.py:1]: ok

File: api/cameras.py (hunk 19 above)
Comment (pgrill) [reply api/cameras.py]: Removed

**Review by mats-claassen — APPROVED** [review]

File: README.md [hunk 20]
```diff
@@ -239,13 +239,12 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
-  - `ScreenshotS3Bucket`: Configures the S3 Bucket used to store the screenshot.
   - `ScreenshotsDirectory`: Configures the folder dedicated to storing all the images (for example the heatmap reports) generated by the processor. We recommend to set this folder to a mounted directory (such as */repo/data/processor/static/screenshots*)
   - `DashboardURL`: Sets the url where the frontend is running. Unless you are using a custom domain, you should keep this value as https://beta.lanthorn.ai/.
   - `SlackChannel`: Configures the slack channel used by the notifications. The chosen slack channel must exist in the configured workspace.
   - `OccupancyAlertsMinInterval`:  Sets the desired interval (in seconds) between occupancy alerts.
   - `MaxThreadRestarts`: Defines the number of restarts allowed per thread.
+  - `HeatmapResolution`: Sets the resolution used by the heatmap report.
 
 - `[Api]`
   - `Host`: Configures the host IP of the processor's API (inside docker). We recommend don't change that value and keep it as *0.0.0.0*.
```
Comment (renzodgc) [inline README.md:250]: `SSLKeyFile` should say:

> If you generate it following the steps defined in this Readme you should put */repo/certs/<your_ip>.key*

Could we include this small change in this PR?

Comment (pgrill) [reply README.md:250]: Done

**Review by mats-claassen — APPROVED** [review]

File: libs/loggers/web_hook_logger.py [hunk 21]
```diff
@@ -29,3 +29,7 @@ def log_objects(self, objects, violating_objects, violating_objects_index_list,
             requests.put(self.web_hook_endpoint, data=request_data)
         except ConnectionError:
             logger.error(f"Connection with endpoint {self.web_hook_endpoint} can't be established")
+
+    def update(self, cv_image, objects, post_processing_data, fps):
+        if self.web_hook_endpoint:
+            super().update(cv_image, objects, post_processing_data, fps)
```
Comment (renzodgc) [inline libs/loggers/web_hook_logger.py:35]: Please add EOL
```suggestion
            super().update(cv_image, objects, post_processing_data, fps)

```

File: libs/utils/loggers.py [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+        if logger_section["Name"] == "file_system_logger":
+            return logger_section["LogDirectory"]
+
+
+def get_source_logging_interval(config):
+    loggers_names = [x for x in config.get_sections() if x.startswith("SourceLogger_")]
+    for l_name in loggers_names:
+        logger_section = config.get_section_dict(l_name)
+        if logger_section["Name"] == "file_system_logger":
+            return float(logger_section["TimeInterval"])
```
Comment (renzodgc) [inline libs/utils/loggers.py:14]:
```suggestion
            return float(logger_section["TimeInterval"])

```

**Review by renzodgc — APPROVED** [review]

**Pull request merged**